File: src/quant/publisher.cpp
#include "quant/publisher.hpp"
#include "util/crypto_hmac.hpp"
#include "util/journal.hpp"
#include "config/config.hpp"
#include <memory>
#include <string>
#include <nlohmann/json.hpp>
#include <chrono>

#if defined(ENABLE_QUANT)
#include <zmq.hpp>
#endif

namespace quant {

using nlohmann::json;

struct NullPublisher : public Publisher {
  void publish_purchase(const journal::Txn&, int, int, const std::string&) override {}
  void publish_drawer_balance(const std::string&, int64_t, int64_t, int64_t, int64_t, const std::string&) override {}
};

#if defined(ENABLE_QUANT)
class ZmqPublisher : public Publisher {
public:
  explicit ZmqPublisher(const Options &opt)
    : opt_(opt), ctx_(1), sock_(ctx_, zmq::socket_type::pub) {
    sock_.connect(opt_.endpoint);
  }

  void publish_purchase(const journal::Txn &t, int price_cents, int deposit_cents,
                        const std::string &idem_key) override {
    json j;
    j["type"] = "purchase";
    j["version"] = 1;
    j["client_id"] = opt_.client_id;
    j["txn_id"] = t.id;
    j["idem_key"] = idem_key;
    // epoch milliseconds
    auto now_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                    std::chrono::system_clock::now().time_since_epoch()).count();
    j["ts_ms"] = now_ms;
    j["amount"] = {
      {"price_cents", price_cents},
      {"deposit_cents", deposit_cents},
      {"change_cents", t.change}
    };

    std::string payload = j.dump();
    if (!opt_.hmac_key_hex.empty()) {
      j["sig"] = util::hmac_sha256_hex(opt_.hmac_key_hex, payload);
      payload = j.dump();
    }

    // Send as [topic][payload]
    zmq::message_t topic(opt_.topic.begin(), opt_.topic.end());
    zmq::message_t body(payload.begin(), payload.end());
    (void)sock_.send(topic, zmq::send_flags::sndmore | zmq::send_flags::dontwait);
    (void)sock_.send(body, zmq::send_flags::dontwait);
  }

  void publish_drawer_balance(const std::string &account_id,
                              int64_t drawer_cents,
                              int64_t pending_change_cents,
                              int64_t reserve_floor_cents,
                              int64_t delta_cents,
                              const std::string &idem_key) override {
    json j;
    j["type"] = "drawer_balance";
    j["version"] = 1;
    j["client_id"] = opt_.client_id;
    j["account_id"] = account_id;
    auto now_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                    std::chrono::system_clock::now().time_since_epoch()).count();
    j["ts_ms"] = now_ms;
    j["currency"] = "USD"; // default; can be extended via config later
    j["balance_cents"] = drawer_cents;
    j["delta_cents"] = delta_cents;
    j["idem_key"] = idem_key;
    j["amounts"] = {
      {"pending_change_cents", pending_change_cents},
      {"reserve_floor_cents", reserve_floor_cents}
    };

    std::string payload = j.dump();
    if (!opt_.hmac_key_hex.empty()) {
      j["sig"] = util::hmac_sha256_hex(opt_.hmac_key_hex, payload);
      payload = j.dump();
    }

    zmq::message_t topic(opt_.topic.begin(), opt_.topic.end());
    zmq::message_t body(payload.begin(), payload.end());
    (void)sock_.send(topic, zmq::send_flags::sndmore | zmq::send_flags::dontwait);
    (void)sock_.send(body, zmq::send_flags::dontwait);
  }

private:
  Options opt_;
  zmq::context_t ctx_;
  zmq::socket_t sock_;
};
#endif

std::shared_ptr<Publisher> Publisher::create(const Options &opt) {
#if defined(ENABLE_QUANT)
  if (!opt.endpoint.empty() && !opt.topic.empty()) {
    try { return std::make_shared<ZmqPublisher>(opt); }
    catch (...) { /* fallthrough to null */ }
  }
#endif
  return std::make_shared<NullPublisher>();
}

Publisher::Options Publisher::from_config(const cfg::Quant &q) {
  Options o;
  o.endpoint = q.endpoint;
  o.topic = q.topic;
  o.client_id = q.client_id;
  o.hmac_key_hex = q.hmac_key_hex;
  o.heartbeat_seconds = q.heartbeat_seconds;
  return o;
}

} // namespace quant

File: integrations/pos/square/webhook_service/app/publisher.py
import os
import time
import hmac
import json
import hashlib
import zmq
from typing import Optional

class ZmqBalancePublisher:
    def __init__(self):
        self.endpoint = os.environ.get("ZMQ_ENDPOINT", "tcp://localhost:5555")
        self.topic = os.environ.get("ZMQ_TOPIC", "register/stream")
        self.client_id = os.environ.get("CLIENT_ID", "REG-CLIENT")
        self.hmac_key_hex = os.environ.get("HMAC_KEY_HEX", "").strip()
        self.currency_default = os.environ.get("CURRENCY", "USD")
        self.reserve_floor_cents = int(os.environ.get("DEFAULT_RESERVE_FLOOR_CENTS", "2000"))
        self._ctx = zmq.Context.instance()
        self._sock = self._ctx.socket(zmq.PUB)
        self._sock.connect(self.endpoint)

    def _hmac(self, payload: str) -> Optional[str]:
        if not self.hmac_key_hex:
            return None
        key = bytes.fromhex(self.hmac_key_hex)
        sig = hmac.new(key, payload.encode("utf-8"), hashlib.sha256).hexdigest()
        return sig

    def publish_drawer_balance(
        self,
        account_id: str,
        balance_cents: int,
        delta_cents: int,
        idem_key: str,
        pending_change_cents: int = 0,
        reserve_floor_cents: Optional[int] = None,
        currency: Optional[str] = None,
    ) -> None:
        j = {
            "type": "drawer_balance",
            "version": 1,
            "client_id": self.client_id,
            "account_id": account_id,
            "ts_ms": int(time.time() * 1000),
            "currency": currency or self.currency_default,
            "balance_cents": int(balance_cents),
            "delta_cents": int(delta_cents),
            "idem_key": idem_key,
            "amounts": {
                "pending_change_cents": int(pending_change_cents),
                "reserve_floor_cents": int(reserve_floor_cents if reserve_floor_cents is not None else self.reserve_floor_cents),
            },
        }
        payload = json.dumps(j, separators=(",", ":"))
        sig = self._hmac(payload)
        if sig:
            j["sig"] = sig
            payload = json.dumps(j, separators=(",", ":"))
        # send multipart [topic][payload]
        self._sock.send_multipart([self.topic.encode("utf-8"), payload.encode("utf-8")], flags=zmq.DONTWAIT)
